pkg/modules/auth/openid/openid.go

```diff
@@ -288,6 +289,33 @@ func getOrCreateUser(s *xorm.Session, cl *claims, provider *Provider, idToken *o
 	return
 }
 
+// mergeClaims combines claims from token and userinfo based on the ForceUserInfo setting
+// cl represents the claims from the token, cl2 represents the claims from userinfo
+func mergeClaims(cl *claims, cl2 *claims, providerName string, forceUserInfo bool) error {
+	if (forceUserInfo && cl2.Email != "") || cl.Email == "" {
+		cl.Email = cl2.Email
+	}
+
+	if (forceUserInfo && cl2.Name != "") || cl.Name == "" {
+		cl.Name = cl2.Name
+	}
+
+	if (forceUserInfo && cl2.PreferredUsername != "") || cl.PreferredUsername == "" {
+		cl.PreferredUsername = cl2.PreferredUsername
+	}
+
+	if cl.PreferredUsername == "" && cl2.Nickname != "" {
+		cl.PreferredUsername = cl2.Nickname
+	}
+
+	if cl.Email == "" {
+		log.Errorf("Claim does not contain an email address for provider %s", providerName)
```

Comment:
Unrelated, but we can get rid of this line and thus the `providerName` parameter

Reply:
Done, the `providerName` parameter is removed and the error logging is moved back to `getCliams`
```go
		if err != nil {
			if user.IsErrNoEmailProvided(err) {
				log.Errorf("Claim does not contain an email address for provider %s", provider.Name)
			}
```